Tests/SwiftFusionTests/Inference/PPCATrackingFactorTests.swift

```diff
@@ -0,0 +1,87 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import TensorFlow
+import XCTest
+
+import PenguinStructures
+@testable import SwiftFusion
+
+class PPCATrackingFactorTests: XCTestCase {
+  /// Tests that the custom `linearized(at:)` method produces the same results at autodiff.
+  func testLinearizedValue() {
+    // Initialize `factor` with random fields.
+    let image = Tensor<Double>(randomNormal: [500, 500, 1])
+    let W = Tensor<Double>(
+      randomNormal: PPCATrackingFactor.Patch.shape + [PPCATrackingFactor.V1.dimension])
+    let mu = PPCATrackingFactor.Patch(
+      Tensor<Double>(randomNormal: PPCATrackingFactor.Patch.shape))
+    let poseID = TypedID<Pose2>(0)
+    let latentID = TypedID<Vector5>(0)
+    let factor = PPCATrackingFactor(poseID, latentID, measurement: image, W: W, mu: mu)
+
+    for _ in 0..<2 {
+      let linearizationPoint = Tuple2(
+        Pose2(randomWithCovariance: eye(rowCount: 3)),
+        Vector5(flatTensor: Tensor(randomNormal: [5])))
+
+      typealias Variables = PPCATrackingFactor.Variables.TangentVector
+      typealias ErrorVector = PPCATrackingFactor.ErrorVector
+
+      let autodiff = JacobianFactor<Array<Variables>, ErrorVector>(
+        linearizing: factor, at: linearizationPoint)
+      let custom = factor.linearized(at: linearizationPoint)
+
+      XCTAssertEqual(
+        custom.errorVector(at: Variables.zero), autodiff.errorVector(at: Variables.zero))
+
+      for _ in 0..<10 {
+        let v = Variables(flatTensor: Tensor(randomNormal: [Variables.dimension]))
+        assertEqual(
+          custom.errorVector_linearComponent(v).tensor,
+          autodiff.errorVector_linearComponent(v).tensor,
+          accuracy: 1e-6)
+      }
+
+      for _ in 0..<10 {
+        let e = ErrorVector(Tensor(randomNormal: ErrorVector.shape))
+        assertEqual(
+          custom.errorVector_linearComponent_adjoint(e).flatTensor,
+          autodiff.errorVector_linearComponent_adjoint(e).flatTensor,
+          accuracy: 1e-6)
+      }
+    }
+  }
+
+  /// Tests that factor graphs with a `PPCATrackingFactor`s linearize them using the custom
+  /// linearization.
+  func testFactorGraphUsesCustomLinearized() {
+    let image = Tensor<Double>(randomNormal: [500, 500, 1])
+    let W = Tensor<Double>(
+      randomNormal: PPCATrackingFactor.Patch.shape + [PPCATrackingFactor.V1.dimension])
+    let mu = PPCATrackingFactor.Patch(
+      Tensor<Double>(randomNormal: PPCATrackingFactor.Patch.shape))
+
+    var x = VariableAssignments()
+    let poseId = x.store(Pose2(100, 100, 0))
+    let latentId = x.store(Vector5.zero)
+
+    var fg = FactorGraph()
+    fg.store(PPCATrackingFactor(poseId, latentId, measurement: image, W: W, mu: mu))
+
+    let gfg = fg.linearized(at: x)
+    XCTAssertNotNil(gfg.storage.first!.value[elementType: Type<LinearizedPPCATrackingFactor>()])
```

Comment:
if you say so ! No idea what's going on here (yet). Comment warranted

Reply:
I think this is an indicator that a higher level API on gfg is needed, encapsulating this test.  Though the test looks awfully specific.  GFG shouldn't be exposing that dictionary in the long run.

Frank, this is testing that the first (key,value) pair in the dictionary underlying gfg has a value that is an `AnyArrayBuffer` containing `LinearizedPPCATrackingFactor` elements.